spec/html_preprocessor_spec.rb

```diff
@@ -0,0 +1,71 @@
+    end
+
+    it "correctly parses protocol-relative url with single quotes" do
+      @body = %{<link href='//fonts.googleapis.com/css?family=Open+Sans:400,600' rel='stylesheet' type='text/css'>}
+      body = @preprocessor.process @body, @root_url, @protocol
+      expect(body).to eq("<link href='http://fonts.googleapis.com/css?family=Open+Sans:400,600' rel='stylesheet' type='text/css'>")
+    end
+
+    it "correctly parses protocol-relative url with double quotes" do
+      @body = %{<link href="//fonts.googleapis.com/css?family=Open+Sans:400,600" rel='stylesheet' type='text/css'>}
+      body = @preprocessor.process @body, @root_url, @protocol
+      expect(body).to eq("<link href=\"http://fonts.googleapis.com/css?family=Open+Sans:400,600\" rel='stylesheet' type='text/css'>")
+    end
+
+    it "correctly parses multiple tags where first one is root url" do
+      @body = %{<a href='/'><img src='/logo.jpg' ></a>}
+      body = @preprocessor.process @body, @root_url, @protocol
+      expect(body).to eq "<a href='http://example.com/'><img src='http://example.com/logo.jpg' ></a>"
+    end
+
+    it "returns the body even if there are no valid substitutions found" do
+      @body = "NO MATCH"
+      body = @preprocessor.process @body, @root_url, @protocol
+      expect(body).to eq("NO MATCH")
+    end
+
+    context 'when root_url is nil' do
+      it "returns the body safely, without interpolating" do
+        @body = %{<link href='//fonts.googleapis.com/css?family=Open+Sans:400,600' rel='stylesheet' type='text/css'><a href='/'><img src='/logo.jpg'></a>}
+        body = @preprocessor.process @body, nil, @protocol
+        expect(body).to eq(%{<link href='http://fonts.googleapis.com/css?family=Open+Sans:400,600' rel='stylesheet' type='text/css'><a href='/'><img src='/logo.jpg'></a>})
+      end
+    end
+
+    context 'when protocol is nil' do
+      it "returns the body safely, without interpolating" do
+        @body = %{<link href='//fonts.googleapis.com/css?family=Open+Sans:400,600' rel='stylesheet' type='text/css'><a href='/'><img src='/logo.jpg'></a>}
+        body = @preprocessor.process @body, @root_url, nil
+        expect(body).to eq(%{<link href='//fonts.googleapis.com/css?family=Open+Sans:400,600' rel='stylesheet' type='text/css'><a href='http://example.com/'><img src='http://example.com/logo.jpg'></a>})
+      end
+    end
+
+    context 'when root_url and protocol are both nil' do
+      it "returns the body safely, without interpolating" do
+        @body = %{<link href='//fonts.googleapis.com/css?family=Open+Sans:400,600' rel='stylesheet' type='text/css'><a href='/'><img src='/logo.jpg'></a>}
+        body = @preprocessor.process @body, nil, nil
+        expect(body).to eq(%{<link href='//fonts.googleapis.com/css?family=Open+Sans:400,600' rel='stylesheet' type='text/css'><a href='/'><img src='/logo.jpg'></a>})
```

Comment:
There's really no need for `@body` to be a class var. Can we change it to `original_body`? (I know you just copied this from elsewhere that did that)

Changing that would allow us to make this test easier to grok, by making this:

```
expect(body).to eq original_body
```